File: tests/typesafety/python/the_program.py
# expect: fails
# title: The bill of materials and the invoice, as they stand today


def bom_line(part_no, description, qty, unit, cost):
    return f"{part_no}  {description}  {qty} {unit} at {cost}d"


def product_cost(costs_pounds):
    return sum(costs_pounds)


def invoice_line(description, price_pence, quantity, taxed):
    note = " +tax" if taxed else ""
    return f"{description} x{quantity} at {price_pence}d{note}"


def invoice_total(prices_pounds):
    return sum(prices_pounds)


def apply_discount(total_pounds, rate):
    return total_pounds * (1 - rate)


def advance(status):
    if status == "darft":
        return "sent"
    if status == "sent":
        return "paid"
    return status


def bill(customer_id, product_id):
    return f"invoice {customer_id} for one {product_id}"


def load_bom_line(row):
    return row.split(",")

File: tests/typesafety/typescript/the_program.ts
// expect: fails

export function bomLine(partNo, description, qty, unit, cost) {
  return `${partNo}  ${description}  ${qty} ${unit} at ${cost}d`;
}

export function productCost(costsPounds) {
  return costsPounds.reduce((sum, cost) => sum + cost, 0);
}

export function invoiceLine(description, pricePence, quantity, taxed) {
  const note = taxed ? " +tax" : "";
  return `${description} x${quantity} at ${pricePence}d${note}`;
}

export function invoiceTotal(pricesPounds) {
  return pricesPounds.reduce((sum, price) => sum + price, 0);
}

export function applyDiscount(totalPounds, rate) {
  return totalPounds * (1 - rate);
}

export function advance(status) {
  if (status === "darft") return "sent";
  if (status === "sent") return "paid";
  return status;
}

export function bill(customerId, productId) {
  return `invoice ${customerId} for one ${productId}`;
}

export function loadBomLine(row) {
  return row.split(",");
}
